ykcompile/src/lib.rs

```diff
@@ -169,51 +169,56 @@ impl<TT> TraceCompiler<TT> {
             let base_ty = self.place_ty(&Place::from(p.local)).clone();
-            match &p.projection[0] {
+            match proj {
                 Projection::Field(idx) => match base_ty {
-                    Ty::Struct(sty) => {
-                        let loc = self.local_to_location(p.local)?;
-                        match loc {
-                            Location::Mem(ro) => {
-                                let offs = sty.fields.offsets[usize::try_from(*idx).unwrap()];
-                                Ok(Location::new_mem(
-                                    ro.reg,
-                                    ro.offs + i32::try_from(offs).unwrap(),
-                                ))
-                            }
-                            _ => unreachable!("{:?}", loc),
+                    Ty::Struct(sty) => match curloc {
+                        Location::Mem(ro) => {
+                            let offs = sty.fields.offsets[usize::try_from(*idx).unwrap()];
+                            curloc =
+                                Location::new_mem(ro.reg, ro.offs + i32::try_from(offs).unwrap());
                         }
-                    }
+                        _ => unreachable!("{:?}", curloc),
+                    },
                     Ty::Tuple(tty) => {
                         if Some(p.local) == self.trace_inputs_local {
+                            // FIXME This case should only happen on the first projection.
                             let offs = tty.fields.offsets[usize::try_from(*idx).unwrap()];
-                            Ok(Location::new_mem(RDI.code(), i32::try_from(offs).unwrap()))
+                            curloc = Location::new_mem(RDI.code(), i32::try_from(offs).unwrap());
                         } else {
-                            let loc = self.local_to_location(p.local)?;
-                            match loc {
+                            match curloc {
                                 Location::Mem(ro) => {
                                     let offs = tty.fields.offsets[usize::try_from(*idx).unwrap()];
-                                    Ok(Location::new_mem(
+                                    curloc = Location::new_mem(
                                         ro.reg,
                                         ro.offs + i32::try_from(offs).unwrap(),
-                                    ))
+                                    );
                                 }
-                                _ => unreachable!("{:?}", loc),
+                                _ => unreachable!("{:?}", curloc),
                             }
                         }
                     }
                     _ => todo!("{:?}", base_ty),
                 },
                 Projection::Deref => {
-                    let loc = self.local_to_location(p.local)?;
+                    // FIXME Dereferencing a reference to a copyable struct/tuple copies it to the
```

Comment:
I take it we don't yet have a way to know if something is `Copy`?

Reply:
I don't think so, but I haven't looked into it much, yet. This is for a future PR.